### PR TITLE
Nz/3348 keystate person

### LINKED ISSUE #3348: Add inline validation for Key State Personnel
Add inline validation for Key State Personnel

If there are no items in the list, it should display the alert message when error checking is on

If the Key Personnel has costs directly attributable to this program

*For fields that are required these validations will be triggered by the error check that is currently being tested. #3488
For fields that have a formatting requirement, they can be done inline after the field is complete. 

|Field|Content|Consideration|
|---|---|---|
|Medicaid director > Name|Provide the name of the State Medicaid Director.|This cannot be blank|
|Medicaid director > Email address|Provide the email address of the State Medicaid Director|This cannot be blank. Do we want to validate as a valid email address?|
|Medicaid director > Phone number|Provide a valid phone number for the State Medicaid Director|This cannot be blank. Do we want to validation as a valid phone number?|
|Medicaid office > Address|Provide a mailing street address for the Medicaid office. |This cannot be blank.|
|Medicaid office > City|Provide a city name.|This cannot be blank|
|Medicaid office > Zip code|Provide a zip code.|This cannot be blank|
|Key personnel/Primary Contact|Must have at least one Primary Contact See #3317 for design||
|KP form > Name|Provide a name for the primary point of contact.|Name of individual should not be blank. See #3433|
|KP form > Email|Provide an email address for the primary point of contact|Email should not be blank. See #3433|
|KP form > Role|Provide a role for the primary point of contact.|Role should not be blank. See #3433|
|KP form > Costs with benefits|Provide a cost greater than or equal to $0|Number equal to 0 is OK because they may have a cost in future FFY only See #3433|
|KP form > FTE Allocation|Provide a number less than or equal to 1. Provide a number greater than or equal to 0.0001| See #3433|
|KP form 2 > Name|Provide a name for the additional key personnel.|Name of individual should not be blank. See #3433|
|KP form 2> Email|Provide an email address for the additional key personnel.|Email should not be blank. See #3433|
|KP form 2> Role|Provide a role for the additional key personnel.|Role should not be blank. See #3433|

**Mockup:**. (How error text would fall, underneath the field.)

### This task is done when...

- [ ] Medicaid director form has inline validation
- [ ] Medicaid office form has inline validation
- [ ] Key Personnel form has inline validation

### PR DESCRIPTION
**Pull request by Sun-Mountain**
Resolves #3348 

### Description

Add inline validation for Key State Personnel to show when Admin Check is enabled.

### Signifcant changes or possible side effects

There shouldn't be any impact to existing functionality or new side effects introduced

### Automated test cases written

- Rewrote Jest test for Key State Medicaid Director and Office.
- Rewrote Cypress test to check for validation errors

### Steps to manually verify this change

1. Go to Key State Personnel
2. Make sure empty required fields show the correct validation error
3. When filled, the errors go away

### This pull request is ready to review when

- [x] Automated tests are updated (and all tests are passing)
- [x] Pull request has been labeled, if applicable with feature, content, bug, tests, refactor
- [ ] Associated OpenAPI documentation has been updated
- [ ] The experience passes a basic manual accessibility audit (keyboard nav, screenreader, text scaling) OR an exemption is documented

### This pull request can be merged when

- [x] Code has been reviewed by someone other than the original author
- [ ] QA has verified the accessibility and functionality related to the change
- [ ] Design has approved the experience
- [ ] Product has approved the experience

### CODE REVIEW

File: api/package.json [hunk 1]
```diff
@@ -68,7 +68,7 @@
     "newrelic": "8.11.1",
     "pg": "8.7.3",
     "sanitize-html": "2.7.0",
-    "sharp": "0.30.5",
+    "sharp": ">=0.30.5",
```
Comment (thetif) [inline api/package.json:71]: can you set this to specifically 0.30.5? we are trying to lock down all the dependencies to specific versions

File: web/src/pages/apd/key-state-personnel/ApdStateProfileMedicaidOffice.js [hunk 2]
```diff
@@ -19,6 +21,41 @@ import { selectKeyStatePersonnel } from '../../../redux/selectors/apd.selectors'
 import { selectState } from '../../../redux/reducers/user';
 import { STATES } from '../../../util';
 
+import Joi from 'joi';
+
+const keyMedicaidSchema = Joi.object({
```
Comment (thetif) [inline web/src/pages/apd/key-state-personnel/ApdStateProfileMedicaidOffice.js:26]: could you put this in the new common/schema file, you can see how it's referenced in ContractorResourcesForm

File: web/src/pages/apd/key-state-personnel/ApdStateProfileMedicaidOffice.js [hunk 3]
```diff
@@ -32,87 +69,210 @@ const ApdStateProfile = ({
...
+        address1: medicaidOffice.address1,
+        address2: medicaidOffice.address2,
+        city: medicaidOffice.city,
+        zip: medicaidOffice.zip
+      }
+    },
+    resolver: joiResolver(keyMedicaidSchema)
+  });
+
+  console.log({adminCheck});
```
Comment (thetif) [inline web/src/pages/apd/key-state-personnel/ApdStateProfileMedicaidOffice.js:98]: remove console.log

File: web/src/static/schemas/medicaidDirector.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
+import Joi from 'joi';
```
Comment (thetif) [inline web/src/static/schemas/medicaidDirector.js:1]: can you move this to common/schema?

**Review by thetif — CHANGES_REQUESTED** [review]
looks good, just a couple of changes needed

File: web/src/pages/apd/key-state-personnel/ApdStateProfileMedicaidOffice.js (hunk 2 above)
Comment (Sun-Mountain) [reply web/src/pages/apd/key-state-personnel/ApdStateProfileMedicaidOffice.js:26]: Yes! I knew I forgot something.

**Review by thetif — APPROVED** [review]

**Comment by mirano-darren** [comment]
Nice work! Found two small things
_Note: Admin check is turned on by default for testing purposes_
| Given  | When | Then | Result |
| ------------- | ------------- | ------------- | ------------- |
| State Director and Office Address fields | fields are blank/filled out | if blank: error message appears; if filled out: error message disappears | PASS ✅ |
| Primary APD PoC/Additional Key Personnel form  | fields are blank/filled out | if blank: error message appears; if filled out: error message disappears | PASS ✅ |
| FTE Allocation field  | a number greater than 1 is entered | the number is considered valid; according to the error message only numbers less than or equal to one are valid. `Provide a number less than or equal to 1. Provide a number greater than or equal to 0.0001` | FAIL ❌ (1)|
| FTE Allocation field  | the field is filled out, then deleted. triggering the validation message again | the old validation error message appears | FAIL ❌ (2)|

1:
image
image

2:
image
image
image

**Comment by mirano-darren** [comment]
Looks good to me!
| Given  | When | Then | Result |
| ------------- | ------------- | ------------- | ------------- |
| State Director and Office Address fields | fields are blank/filled out | if blank: error message appears; if filled out: error message disappears | PASS ✅ |
| Primary APD PoC/Additional Key Personnel form  | fields are blank/filled out | if blank: error message appears; if filled out: error message disappears | PASS ✅ |
| FTE Allocation field  | a number greater than 1 is entered | the number is considered invalid | PASS ✅ |
| FTE Allocation field  | the field is filled out, then deleted. triggering the validation message again | the correct validation error message appears | PASS ✅ |

**Review by beparticular — COMMENTED** [review]
@jeromeleecms @thetif 
Re: state - were we going to automatically assign the state on the State Medicaid Director state field? Re: in this test I'm seeing the account as Alaska and the state field defaulting to the top (Alabama)

**Comment by beparticular** [comment]
Additional comment - I'm seeing this happen where I select an option, but it's noting a validation issue

Screen Shot 2022-06-29 at 10 47 51 AM

If you select "no" first before trying to enter any other fields, it doesn't show the validation error until you add more info in the fields above it, so it may look like nothing is wrong if you choose no before filling out or tabbing through the other fields. Happens in both Primary POC and in AKP.

**Comment by mirano-darren** [comment]
> Additional comment - I'm seeing this happen where I select an option, but it's noting a validation issue
> 
> Screen Shot 2022-06-29 at 10 47 51 AM
> 
> If you select "no" first before trying to enter any other fields, it doesn't show the validation error until you add more info in the fields above it, so it may look like nothing is wrong if you choose no before filling out or tabbing through the other fields. Happens in both Primary POC and in AKP.

hmmm I wasn't able to recreate this. Could you list out the exact steps you did? @beparticular

**Review by jeromeleecms — CHANGES_REQUESTED** [review]
For the FTE allocation, the error we should be checking for is:
1) Number <= 1
2) 0 is a valid number for FTE allocation (because they may have a cost in one FFY, but not in another as we reasoned out for the "Cost with benefits" portion the subform section)
3) Error message can be: "Provide a number less than or equal to 1. Provide a number greater than or equal to 0."

I see that's not how it's written in the original issue, so this change request is associated with me missing this in writeup for the acceptance criteria.

**Review by tbolt — APPROVED** [review]

**Pull request merged**